File: backend/app/schemas/district.py
"""District and Block Pydantic schemas."""
import uuid
from datetime import datetime
from typing import Optional, Any, List, TYPE_CHECKING
from pydantic import BaseModel

if TYPE_CHECKING:
    from app.schemas.monitoring_station import MonitoringStationOverview


class DistrictBase(BaseModel):
    name: str
    avg_porosity: Optional[float] = None
    avg_hydraulic_conductivity: Optional[float] = None
    vulnerability_index: Optional[float] = None


class DistrictCreate(DistrictBase):
    geometry: Optional[Any] = None  # GeoJSON FeatureCollection or dict


class DistrictUpdate(BaseModel):
    name: Optional[str] = None
    geometry: Optional[Any] = None
    avg_porosity: Optional[float] = None
    avg_hydraulic_conductivity: Optional[float] = None
    vulnerability_index: Optional[float] = None


class DistrictResponse(DistrictBase):
    id: uuid.UUID
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ── Blocks ────────────────────────────────────────────────────────

class BlockBase(BaseModel):
    name: str
    district_id: uuid.UUID
    aquifer_distribution: Optional[dict] = None
    avg_porosity: Optional[float] = None
    avg_permeability: Optional[float] = None


class BlockCreate(BlockBase):
    geometry: Optional[Any] = None


class BlockUpdate(BaseModel):
    name: Optional[str] = None
    geometry: Optional[Any] = None
    aquifer_distribution: Optional[dict] = None
    avg_porosity: Optional[float] = None
    avg_permeability: Optional[float] = None


class BlockResponse(BlockBase):
    id: uuid.UUID
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class BlockDetailResponse(BlockResponse):
    """Extended block response that includes a summary of linked monitoring stations."""
    monitoring_stations: List[Any] = []  # List[MonitoringStationOverview] – avoid circular import

    model_config = {"from_attributes": True}

File: frontend/src/types/district.ts
import type { GeoJSONGeometry } from './common';

export interface District {
    id: string;
    name: string;
    avg_porosity: number | null;
    avg_hydraulic_conductivity: number | null;
    vulnerability_index: number | null;
    geometry: GeoJSONGeometry | null;
    created_at: string;
    updated_at: string;
}

export interface DistrictCreate {
    name: string;
    avg_porosity?: number;
    avg_hydraulic_conductivity?: number;
    vulnerability_index?: number;
    geometry?: GeoJSONGeometry;
}

export interface DistrictUpdate extends Partial<DistrictCreate> { }

export interface Block {
    id: string;
    name: string;
    district_id: string;
    aquifer_distribution: Record<string, number> | null;
    avg_porosity: number | null;
    avg_permeability: number | null;
    geometry: GeoJSONGeometry | null;
    created_at: string;
    updated_at: string;
}

export interface BlockCreate {
    name: string;
    district_id: string;
    aquifer_distribution?: Record<string, number>;
    avg_porosity?: number;
    avg_permeability?: number;
    geometry?: GeoJSONGeometry;
}

export interface BlockUpdate extends Partial<Omit<BlockCreate, 'district_id'>> { }
